### PR TITLE
Rewrite understanding client-side tools section

### LINKED ISSUE #12495: [Tools] fails to build; problem with postcss-preset-env versions
MDN URL: https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Introducing_complete_toolchain

#### What information was incorrect, unhelpful, or incomplete?
There is an issue with instructions that result in the project failing to build/transform. I have troubleshooted the issue to what appears to be a problem with postcss-preset-env versions.

#### Specific section or headline?
The instructions throughout this chapter lead to the resultant error. Need to run the fix prescribed in linked comment before the **Running the transformation** step.

#### What did you expect to see?
Not encounter errors, build succeeds and resultant page is accessible.

#### Did you test this? If so, how?
Yes. See this [comment](https://discourse.mozilla.org/t/help-needed-with-introducing-a-complete-toolchain/88832/16?u=horusofoz) in a thread on the MDN Discourse forum to see the troubleshooting and workaround/fix. I expect those familiar with npm and relevant packages will find a more exact fix but that referenced does work in the interim.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/learn/tools_and_testing/understanding_client-side_tools/introducing_complete_toolchain`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Introducing_complete_toolchain
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/tools_and_testing/understanding_client-side_tools/introducing_complete_toolchain/index.md
* Last commit: https://github.com/mdn/content/commit/07453ac2b659a397b0fee921fa71db5488e9ca62
* Document last modified: 2021-10-08T14:43:33.000Z

</details>

### LINKED ISSUE #17278: [Tools] Parcel example should use v2
### MDN URL

https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Package_management

### What specific section or headline is this issue about?

Having fun with Parcel

### What information was incorrect, unhelpful, or incomplete?

Command:
npm install parcel-bundler
result:
npm WARN deprecated parcel-bundler@1.12.5: Parcel v1 is no longer maintained. Please migrate to v2, which is published under the 'parcel' package. See https://v2.parceljs.org/getting-started/migration for details.

Command:
npm install parcel
result:
Ok.

Command:
npx parcel index.html
result:
× Build failed.
@parcel/core: Failed to resolve 'date-fns' from './index.js'

### What did you expect to see?

I want to see the working example. :)

### Do you have any supporting links, references, or citations?

_No response_

### Do you have anything more you want to share?

_No response_

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/learn/tools_and_testing/understanding_client-side_tools/package_management`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Package_management
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md
* Last commit: https://github.com/mdn/content/commit/73ea2b0f8ebfab23630b86e64b14a7b1f7577679
* Document last modified: 2022-05-13T04:41:50.000Z

</details>

### LINKED ISSUE #18616: [Tools] using npx does not guarantee correct version or library
### MDN URL

https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Introducing_complete_toolchain

### What specific section or headline is this issue about?

Running the transformation

### What information was incorrect, unhelpful, or incomplete?

Running parcel with `npx` may be use different version than locally installed depending on the npm version and does user has the package installed globally / locally / not at all. This can lead to incompatible configuration / versions. As the parcel-bundler is already installed locally, it would make more sense to use that version by packege.json scripts, like parcel is instructing https://en.parceljs.org/getting_started.html#adding-parcel-to-your-project 

Here is at least one issue about which version npx is using: https://github.com/npm/cli/issues/2329 

### What did you expect to see?

Instruct user to add dev / build script to run parcel.

### Do you have any supporting links, references, or citations?

_No response_

### Do you have anything more you want to share?

_No response_

### LINKED ISSUE #28080: [Tools] Content about Parcel is deprecated and troublesome
### MDN URL

https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Package_management

### What specific section or headline is this issue about?

Having fun with Parcel & Building our code for production

### What information was incorrect, unhelpful, or incomplete?

I list the three commands I fail to follow along
### npm install parcel-bundler 
This package has been [deprecated](https://www.npmjs.com/package/parcel-bundler), should just install [parcel](https://www.npmjs.com/package/parcel)
Note that I would love to use the deprecated version but it can't even be installed anymore

I know the article would mention the version issue later but it should be fixed in this early part unless the author want  the audience to find out themselves

### parcel index.html
TL;DR --- Parcel failed to auto install 
* parcel didn't auto install the date-fns package but reporting build failed instead
    *  @parcel/transformer-js: Browser scripts cannot have imports or exports.

* Following the suggestion I added type="module" to the script tag. The problem was not solved while other errors occurred 
   * @parcel/core: Failed to resolve 'date-fns' from './index.js'
   * @parcel/resolver-default: Cannot find module 'date-fns'

* To find out whether parcel can auto install packages, I tried add a file named "style.sass" and include it in a link tag in index.html, parcel did actually install sass for me with 
  ```
  "devDependencies": {
    "@parcel/transformer-sass": "^2.9.3"
  }
  ```
* I thought there was something wrong with date-fns, but then I tried another popular package [dayjs](https://www.npmjs.com/package/dayjs), it failed with the same error messages
   * @parcel/core: Failed to resolve 'dayjs' from './index.js'
   * @parcel/resolver-default: Cannot find module 'dayjs'

While I fail to solve, I think I should at least post this issue.

### parcel build index.html
error: 
* @parcel/namer-default: Target "main" declares an output file path of "index.js" which does
	not match the compiled bundle type "html".

	solution: [remove the "main" field in package.json](https://stackoverflow.com/questions/68884914/where-should-the-path-from-main-in-package-json-point-to)

### Conclusion
I think this part should be updated to some extent since it's hard to follow along, and I can not even figure out the solution.

My environment: 
Windows 11 / Ubuntu 22.04.2 LTS run on WSL
node: v18.17.0 / v20.4.0
parcel: 2.9.3

	

### What did you expect to see?

Commands that can be followed along like all other tutorials
To be specific: 
1. Viable parcel version
2. Successful package auto installation
3. Successfully parcel build with default command and npm init settings

### Do you have any supporting links, references, or citations?

_No response_

### Do you have anything more you want to share?

Pasted image 20230720174909
Pasted image 20230720175107
Pasted image 20230720175132
Pasted image 20230720175452

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/learn/tools_and_testing/understanding_client-side_tools/package_management`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Package_management
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md
* Last commit: https://github.com/mdn/content/commit/acfe8c9f1f4145f77653a2bc64a9744b001358dc
* Document last modified: 2023-07-07T07:19:19.000Z

</details>

### LINKED ISSUE #28100: [Tools] Deprecated packages and links, issue about parcel auto installation, wrong path in /src/index.css
### MDN URL

https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Introducing_complete_toolchain

### What specific section or headline is this issue about?

Tools used in our toolchain; Installing our tools; Build and transformation tools; Running the transformation

### What information was incorrect, unhelpful, or incomplete?

1. [JSX introduction site](https://legacy.reactjs.org/docs/introducing-jsx.html) deprecated, new site at [Writing Markup with JSX](https://react.dev/learn/writing-markup-with-jsx) & [JavaScript in JSX with Curly Braces](https://react.dev/learn/javascript-in-jsx-with-curly-braces)
2. [babel-eslint](https://www.npmjs.com/package/babel-eslint) deprecated, now [@babel/eslint-parser](https://www.npmjs.com/package/@babel/eslint-parser)
3. [parcel-bundler](https://www.npmjs.com/package/parcel-bundler) deprecated. Can't even npm install it now
4. command ```npx parcel src/index.html``` failed
 #### TL;DR 
* Install the packages which should be auto installed by parcel on your own 
* Change the file path in /src/index.css; 

Below is the process of debugging. First run the command ```npx parcel src/index.html``` and an error occurred

>🚨 Build failed.
	@parcel/transformer-js: Expression expected
	image

As the parcel auto installation issue I posted #28080 , parcel didn't auto install react, react-dom, react-async-hook, date-fns, and format-number for me like the tutorial said. 
So I install them on my own and encounter next little problem, the wrong path of svg file 
> image

Changing absolute path '/img/filename.svg' to relative path 'img/filename.svg' solved the issue while next error occurred 
>🚨 Build failed.
Error: Expected content key 37a1035e8011b3a1 to exist

Easy, just delete the cache directory of parcel by running ```rm -r .parcel-cache```
Now parcel build successfully, damn the auto installation!
I need to mention that I also change babel-eslint(now deprecated) to @babal/core and @babel/eslint-parser, not sure if that makes any difference

### What did you expect to see?

1. viable packages
2. parcel build successfully

### Do you have any supporting links, references, or citations?

_No response_

### Do you have anything more you want to share?

My environment:
Windows 11 / Ubuntu 22.04.2 LTS run on WSL
node: v18.17.0 / v20.4.0
parcel: 2.9.3

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/learn/tools_and_testing/understanding_client-side_tools/introducing_complete_toolchain`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Introducing_complete_toolchain
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/tools_and_testing/understanding_client-side_tools/introducing_complete_toolchain/index.md
* Last commit: https://github.com/mdn/content/commit/985ae59480b490c330af2d40f765ddbe31e6269c
* Document last modified: 2023-07-10T16:35:16.000Z

</details>

### LINKED ISSUE #31712: [Tools] The option -r of rmdir not supported on Windows/mac
### MDN URL

https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Command_line

### What specific section or headline is this issue about?

Creating, copying, moving, removing

### What information was incorrect, unhelpful, or incomplete?

The option `-r` of `rmdir` has been not supported on **Window 10**.

### What did you expect to see?

Use `/s` instead.

### Do you have any supporting links, references, or citations?

_No response_

### Do you have anything more you want to share?

_No response_

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/learn/tools_and_testing/understanding_client-side_tools/command_line`
* MDN URL: https://developer.mozilla.org/en-US/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Command_line
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/learn/tools_and_testing/understanding_client-side_tools/command_line/index.md
* Last commit: https://github.com/mdn/content/commit/dfedc6e2f59e7072a9ced35ce4049a906438d74c
* Document last modified: 2023-08-11T00:30:18.000Z

</details>

### PR DESCRIPTION
**Pull request by Josh-Cena**
- Fix https://github.com/mdn/content/issues/31712
- Fix https://github.com/mdn/content/issues/17278
- Fix https://github.com/mdn/content/issues/28080
- Fix https://github.com/mdn/content/issues/12495
- Fix https://github.com/mdn/content/issues/18616
- Fix https://github.com/mdn/content/issues/28100
- Remove Parcel because I don't think anyone here is familiar with it and it does a lot of magic around dependency management which is bad for education purposes
- Self-host the example repo
- Other modernizations around references to toolings

### CODE REVIEW

**Comment by Josh-Cena** [comment]
Content is mostly ready, pending the part of GitHub Pages deployment because I need @bsmth's help

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md [hunk 1]
```diff
@@ -38,70 +38,34 @@ Here's just a few things to consider for this particular project:
...
 - Actually deploying the updated code to a live URL: Or potentially a staging URL so it can be reviewed first.
 
-> **Note:** Cache busting is a new term that we haven't met before in the module. This is the strategy of breaking a browser's own caching mechanism, which forces the browser to download a new copy of your code. Parcel (and indeed many other tools) will generate filenames that are unique to each new build. This unique filename "busts" your browser's cache, thereby making sure the browser downloads the fresh code each time an update is made to the deployed code.
+> **Note:** Cache busting is a new term that we haven't met before in the module. This is the strategy of breaking a browser's own caching mechanism, which forces the browser to download a new copy of your code. Vite (and indeed many other tools) will generate filenames that are unique to each new build. This unique filename "busts" your browser's cache, thereby making sure the browser downloads the fresh code each time an update is made to the deployed code.
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:41]: Is this the first mention of Vite? Anywhere earlier we can introduce it and link to https://vitejs.dev/ ?

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md [hunk 2]
```diff
@@ -38,70 +38,34 @@ Here's just a few things to consider for this particular project:
...
 
 However, we do have to set these steps up, and we'll look at that now.
 
 ## The build process
 
-Again, because we're using Parcel for development, the build option is extremely simple to add. Instead of running the server with `npx parcel src/index.html`, we can run it with `npx parcel build src/index.html` and Parcel will build everything ready for production instead of just running it for development and testing purposes. This includes doing minification and tree-shaking of code, and cache-busting on filenames.
+Again, because we're using Vite for development, the build option is extremely simple to add. As we saw earlier, we already have a custom script `npm run build` that will let Vite build everything ready for production instead of just running it for development and testing purposes. This includes doing minification and tree-shaking of code, and cache-busting on filenames.
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:61]: We have glossary entries for these, want to add them?

* https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking
* https://developer.mozilla.org/en-US/docs/Glossary/Minification

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md [hunk 3]
```diff
@@ -225,92 +155,106 @@ When approaching tests there are a good deal of ways to approach the problem:
...
 
 Remember also that tests are not limited to JavaScript; tests can be run against the rendered DOM, user interactions, CSS, and even how a page looks.
 
-However, for this project we're going to create a small test that will check the third-party NASA data feed and ensure it's in the correct format. If not, the test will fail and will prevent the project from going live. To do anything else would be beyond the scope of this module — testing is a huge subject that really requires its own separate module. We are hoping that this section will at least make you aware of the need for testing, and will plant the seed that inspires you to go and learn more.
+However, for this project we're going to create a small test that will check the GitHub API data is in the correct format. If not, the test will fail and will prevent the project from going live. To do anything else would be beyond the scope of this module — testing is a huge subject that really requires its own separate module. We are hoping that this section will at least make you aware of the need for testing, and will plant the seed that inspires you to go and learn more.
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:162]: Minor typo

```suggestion
However, for this project we're going to create a small test that will check if the GitHub API data is in the correct format. If not, the test will fail and will prevent the project from going live. To do anything else would be beyond the scope of this module — testing is a huge subject that really requires its own separate module. We are hoping that this section will at least make you aware of the need for testing, and will plant the seed that inspires you to go and learn more.
```

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md [hunk 4]
```diff
@@ -225,92 +155,106 @@ When approaching tests there are a good deal of ways to approach the problem:
...
 
-Although the test for this project does not include a test framework, as with all things in the front-end development world, there are a slew of [framework options](https://www.npmjs.com/search?q=keywords%3Atesting).
+The test itself isn't what is important. What is important is how the failure or success is handled. Because we are writing a custom build action already, we can just add a step before the build that runs the test. If the test fails, the build will fail, and the deployment will not happen.
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:164]: Style suggestion to remove "just":

```suggestion
The test itself isn't what is important. What is important is how the failure or success is handled. Because we are writing a custom build action already, we can add a step before the build that runs the test. If the test fails, the build will fail, and the deployment will not happen.
```

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md [hunk 5]
```diff
@@ -225,92 +155,106 @@ When approaching tests there are a good deal of ways to approach the problem:
...
+   ```
+
+2. In your package.json, find your `scripts` member, and update it so that it contains the following test and build commands:
 
    ```json
    "scripts": {
-     …
-     "test": "node tests/*.js",
-     "build": "npm run test && parcel build src/index.html"
+     // …
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:180]: We have "the following test and build commands" but only a test command. Should we include the `build` command?

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md [hunk 6]
```diff
@@ -225,92 +155,106 @@ When approaching tests there are a good deal of ways to approach the problem:
...
 
-Although the test for this project does not include a test framework, as with all things in the front-end development world, there are a slew of [framework options](https://www.npmjs.com/search?q=keywords%3Atesting).
+The test itself isn't what is important. What is important is how the failure or success is handled. Because we are writing a custom build action already, we can just add a step before the build that runs the test. If the test fails, the build will fail, and the deployment will not happen.
 
-The test itself isn't what is important. What is important is how the failure or success is handled. Some deployment platforms will include a specific method for testing as part of their pipeline. Products like GitHub, GitLab, etc., all support running tests against individual commits.
-
-As this project is deploying to Netlify, and Netlify only asks about the build command, we will have to make the tests part of the build. If the test fails, the build fails, and Netlify won't deploy.
+The good news is: because we are using Vite, Vite already offers a good integrated tool for testing: [Vitest](https://vitest.dev/guide/).
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:166]: There's also the built-in test runner in https://nodejs.org/api/test.html, maybe it's worth a mention?

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md [hunk 7]
```diff
@@ -225,92 +155,106 @@ When approaching tests there are a good deal of ways to approach the problem:
...
+         press h to show help, press q to quit
    ```
 
-   The result, if successful, is … nothing. This is considered a success. In general, we only want tests to be noisy if there's something wrong. The test also exited with a special signal that tells the command line that it was successful — an exit signal of 0. If there's a failure the test fails with an exit code of 1 — this is a system-level value that says "something failed".
+   This means the test passed. Like Vite, it will watch for changes and re-run the tests when you save a file. We can just quit by pressing `q`.
 
-   The `npm run test` command will use node to run all the files that are in the tests directory that end with `.js`.
+5. We still need to wire the test to our build action, so it blocks the build if the test fails. Open the `.github/workflows/deploy.yml` file and add the following step:
+
+   TODO: where?
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:221]: The gh-pages workflow is here https://github.com/mdn/client-toolchain-example/blob/main/.github/workflows/github-pages.yml, we can rename it to `.github/workflows/deploy.yml` if you like

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md [hunk 8]
```diff
@@ -225,92 +155,106 @@ When approaching tests there are a good deal of ways to approach the problem:
...
+
+   ```yaml
+   - name: Run tests
+     run: npm run test
+   ```
 
-   In our build script, `npm run test` is called, then you see the string `&&` — this means "if the thing on the left succeeded (exited with zero), then do this thing on the right". So this translates into: if the tests pass, then build the code.
+   This will run the test before the build step. If the test fails, the build will fail, and the deployment will not happen.
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:228]: I think we're repeating the same as on L164 https://github.com/mdn/content/pull/34017/files#diff-53c5b036d0d061dc9a867d10c577da205fc9ffe1749f2ae56294ae9450e31214R164

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md [hunk 9]
```diff
@@ -225,92 +155,106 @@ When approaching tests there are a good deal of ways to approach the problem:
...
+- There are two configuration files that the code quality tools read: `eslint.config.js` and `.prettierrc`.
+- During development, we continue to add dependencies using npm. The Vite development server is running in the background to watch for changes and to automatically build our source.
+- Deployment is handled by pushing our changes to GitHub (on the "main" branch), which triggers a build and deployment using GitHub Actions to publish the project. For our instance this URL is <https://mdn.github.io/client-toolchain-example>; you will have your own unique URL.
+- We also have a simple test that blocks the building and deployment of the site if the GitHub API feed isn't giving us the correct data format.
 
 For those of you wanting a challenge, consider whether you can optimize some part of this toolchain. Some questions to ask yourself:
 
-- Can [images be compressed](https://github.com/ralscha/parcel-plugin-compress) during the build step?
+- Can we extract only the features of plotly.js that we need? This will reduce the JavaScript bundle size.
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:256]: Is this the first mention of plotly? Should we introduce it earlier so readers aren't surprised or guessing what it is?

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md (hunk 1 above)
Comment (bsmth) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:41]: I'm reviewing in the wrong order, scratch that.

Comment (Josh-Cena) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:41]: This is far from the first. It's the last in the sequence and we've been using Vite all along.

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/introducing_complete_toolchain/index.md [hunk 10]
```diff
@@ -66,28 +61,55 @@ However, more complicated software requirements will likely benefit from the usa
 
 For our sample project, we'll be using a toolchain specifically designed to aid our software development and support the technical choices made during the software design phase. We will however be avoiding any superfluous tooling, with the aim of keeping complexity to a minimum.
 
-For example, we _could_ have included a tool to minimize our SVG file sizes during the build. However, this project has only 4 SVG images, which were [manually minified using SVGO](https://www.npmjs.com/package/svgo) before adding them to the project.
+## Checking prerequisites
+
+You should have most of the softwares already if you've been following along with the previous chapters. Here's what you should have before proceeding to the real setup steps. They only need to be done once and you don't need to repeat these again for future projects.
+
+### Creating GitHub account
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/introducing_complete_toolchain/index.md:68]:
```suggestion
### Creating a GitHub account
```

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md (hunk 5 above)
Comment (Josh-Cena) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:180]: The `build` command should already be added previously.

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md (hunk 6 above)
Comment (Josh-Cena) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:166]: The built-in test runner is supposed to be a primitive to be used with other tools, and we are testing JSX here which takes _a lot_ of work to integrate, so overall I don't think it's worth it.

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/command_line/index.md [hunk 11]
```diff
@@ -365,7 +365,7 @@ For this article, we will install Prettier locally, as suggested in the [Prettie
 Once you've installed node, open up the terminal and run the following command to install Prettier:
 
 ```bash
-npm install prettier
+npm install --save-dev prettier
```
[the remark this answers is not in the document]
Comment (bsmth) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/command_line/index.md:368]: We do, resolving

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/overview/index.md [hunk 12]
```diff
@@ -113,6 +117,11 @@ Generally, web development is thought of as three languages: [HTML](/en-US/docs/
    - [TypeScript](https://www.typescriptlang.org/): TypeScript is a superset of JavaScript that offers a bunch of additional features. The TypeScript compiler converts TypeScript code to JavaScript when building for production.
    - Frameworks such as [React](https://react.dev/), [Ember](https://emberjs.com/), and [Vue](https://vuejs.org/): Frameworks provide a lot of functionality for free and allow you to use it via custom syntax built on top of vanilla JavaScript. In the background, the framework's JavaScript code works hard to interpret this custom syntax and render it as a final web app.
 
+3. Optimization. This is provided by _bundlers_, which are tools that get your code ready for production, for example by "tree-shaking" to make sure only the parts of your code libraries that you are actually using are put into your final production code, or "minifying" to remove all the whitespace in your production code, making it as small as possible before it is uploaded to a server. For example:
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/overview/index.md:120]: Same comment about linking to https://developer.mozilla.org/en-US/docs/Glossary/Tree_shaking and https://developer.mozilla.org/en-US/docs/Glossary/Minification if you like

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/overview/index.md [hunk 13]
```diff
@@ -113,6 +117,11 @@ Generally, web development is thought of as three languages: [HTML](/en-US/docs/
...
+
+   - [Webpack](https://webpack.js.org/) has been the most popular bundler for a long time, featuring a huge number of plugins and a powerful configuration system. However, it is also known for being quite complex to set up, and is slow compared to more modern alternatives.
+   - [Vite](https://vitejs.dev/) is a new build tool that is very quickly gaining popularity for its speed, simplicity, and richness of features.
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/overview/index.md:123]: Maybe we can future-proof this a little more:

```suggestion
   - [Vite](https://vitejs.dev/) is a modern build tool that is popular for its speed, simplicity, and richness of features.
```

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md [hunk 14]
```diff
@@ -77,250 +77,218 @@ What is important is that you ensure you've chosen the best registry for you. Ma
...
+We will use [Vite](https://vitejs.dev/) to create a blank website. In the next article, we will expand on the toolchain to include more tools and show you how to deploy the site.
+
+Vite provides some [init templates](https://vitejs.dev/guide/#scaffolding-your-first-vite-project), with all necessary dependencies and configurations, to get you started quickly in a real project. For demonstration, we will configure one from scratch, using the [React template](https://github.com/vitejs/vite/tree/main/packages/create-vite/template-react) as a reference.
 
 ### Setting up the app as an npm package
 
-First of all, create a new directory to store our experimental app in, somewhere sensible that you'll find again. We'll call it parcel-experiment, but you can call it whatever you like:
+First of all, create a new directory to store our experimental app in, somewhere sensible that you'll find again. We'll call it npm-experiment, but you can call it whatever you like:
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md:86]:
```suggestion
First of all, create a new directory to store our experimental app in, somewhere sensible that you'll find again. We'll call it `npm-experiment`, but you can call it whatever you like:
```

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md [hunk 15]
```diff
@@ -77,250 +77,218 @@ What is important is that you ensure you've chosen the best registry for you. Ma
...
   },
-  "author": "Chris Mills",
+  "author": "Your name",
   "license": "ISC"
 }
 ```
 
+We will add two more lines to package.json:
+
+- `"type": "module"`, which causes Node to interpret all .js files as [ES modules](/en-US/docs/Web/JavaScript/Guide/Modules) rather than the old CommonJS modules. It's a generally good habit to get into.
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md:131]: Thoughts on formatting this one?

```suggestion
- `"type": "module"`, which causes Node to interpret all `.js` files as [ES modules](/en-US/docs/Web/JavaScript/Guide/Modules) rather than the old CommonJS modules. It's a generally good habit to get into.
```

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/command_line/index.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Josh-Cena) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/command_line/index.md:368]: This is taken from Prettier's docs and I don't want to explain it further without talking about package.json first, so I just added a forward-looking note.

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md (hunk 9 above)
Comment (Josh-Cena) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:256]: This is the last article in the sequence and we are using plotly in our code.

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md (hunk 15 above)
Comment (Josh-Cena) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md:131]: I'm indifferent. I slightly prefer file names to not be code entities but no strong opinions.

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md [hunk 16]
```diff
@@ -77,250 +77,218 @@ What is important is that you ensure you've chosen the best registry for you. Ma
...
+Now we'll run our newly installed Vite tool. In your terminal, run the following command:
 
 ```bash
- npx parcel index.html
+npx vite
 ```
 
 You should see something like this printed in your terminal:
 
 ```bash
```
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md:208]: Should we use `plain` here?

```suggestion
```plain
```

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md
Comment (bsmth) [inline files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md:1]: We mention in L66:

> **Note:** npm is not the only package manager available. A successful and popular alternative package manager is [Yarn](https://yarnpkg.com/). Yarn resolves the dependencies using a different algorithm that can mean a faster user experience. There are also a number of other emerging clients, such as [pnpm](https://pnpm.js.org/).

I'm not sure this is true anymore in recent years. Are there still benefits to Yarn over npm? We should still mention it because readers are very likely to encounter it in projects.

Also later in the initialize phase, we're running npm and yarn commands sequentially without explaining that you shouldn't mix these in projects, and it's either-or.

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md (hunk 6 above)
Comment (bsmth) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:166]: OK sounds good

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md (hunk 9 above)
Comment (bsmth) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:256]: I started reviewing in the wrong order, ignore 😄

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md (hunk 15 above)
Comment (bsmth) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md]: I'll leave it to you 👍🏻

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md
Comment (Josh-Cena) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md:1]: > I'm not sure this is true anymore in recent years. Are there still benefits to Yarn over npm? We should still mention it because readers are very likely to encounter it in projects.

There are many ways in which Yarn is faster than npm.

- In a monorepo, Yarn offers the best experience with precise hoisting.
- Yarn PnP wraps its own Node runtime that is faster than node_modules resolution.
- Zero-installs allows you to commit node_modules (basically, but compressed of course) and others won't even need to ever install
- Yarn and pnpm both have faster disk operations than npm.

I'm not sure if we should describe in such detail. We just picked tools that work, not saying there's anything worse about other tools.

> Also later in the initialize phase, we're running npm and yarn commands sequentially without explaining that you shouldn't mix these in projects, and it's either-or.

Good idea 👍

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md (hunk 8 above)
Comment (Josh-Cena) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/deployment/index.md:228]: L164 describes the intention/plan. This line describes the effect of this addition. The effect matches the intent but they are still independent.

File: files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md
Comment (bsmth) [reply files/en-us/learn/tools_and_testing/understanding_client-side_tools/package_management/index.md:1]: Super, thanks for the info!

>I'm not sure if we should describe in such detail.

No that's fine, thanks.

**Review by bsmth — APPROVED** [review]
I think we're all good to go. In general, I'm trying to avoid adding screenshots because of the maintenance cost, if you think we can go without them using links instead, that would be great. Otherwise, leaving a +1 on this. Well done 👍🏻

**Comment by Josh-Cena** [comment]
Great! Yeah I think the remaining screenshots are necessary.
